File: src/app/src/request/transcript.rs
use super::payload::extract_content_text;
use crate::storage::model::ToolCallRecord;
use crate::storage::model::TranscriptItem;

pub fn from_agent_messages(messages: &[bend_engine::AgentMessage]) -> Vec<TranscriptItem> {
    messages.iter().map(TranscriptItem::from).collect()
}

pub fn into_agent_messages(items: &[TranscriptItem]) -> Vec<bend_engine::AgentMessage> {
    items.iter().map(bend_engine::AgentMessage::from).collect()
}

impl From<&bend_engine::AgentMessage> for TranscriptItem {
    fn from(message: &bend_engine::AgentMessage) -> Self {
        match message {
            bend_engine::AgentMessage::Llm(bend_engine::Message::User { content, .. }) => {
                let text = extract_content_text(content);
                TranscriptItem::User { text }
            }
            bend_engine::AgentMessage::Llm(bend_engine::Message::Assistant { content, .. }) => {
                let mut text = String::new();
                let mut thinking = None;
                let mut tool_calls = Vec::new();

                for block in content {
                    match block {
                        bend_engine::Content::Text { text: chunk } => {
                            if !text.is_empty() {
                                text.push('\n');
                            }
                            text.push_str(chunk);
                        }
                        bend_engine::Content::Thinking {
                            thinking: chunk, ..
                        } => {
                            thinking = Some(chunk.clone());
                        }
                        bend_engine::Content::ToolCall {
                            id,
                            name,
                            arguments,
                        } => {
                            tool_calls.push(ToolCallRecord {
                                id: id.clone(),
                                name: name.clone(),
                                input: arguments.clone(),
                            });
                        }
                        _ => {}
                    }
                }

                TranscriptItem::Assistant {
                    text,
                    thinking,
                    tool_calls,
                }
            }
            bend_engine::AgentMessage::Llm(bend_engine::Message::ToolResult {
                tool_call_id,
                tool_name,
                content,
                is_error,
                ..
            }) => {
                let text = extract_content_text(content);
                TranscriptItem::ToolResult {
                    tool_call_id: tool_call_id.clone(),
                    tool_name: tool_name.clone(),
                    content: text,
                    is_error: *is_error,
                }
            }
            bend_engine::AgentMessage::Extension(ext) => TranscriptItem::Extension {
                kind: ext.kind.clone(),
                data: ext.data.clone(),
            },
        }
    }
}

impl From<&TranscriptItem> for bend_engine::AgentMessage {
    fn from(item: &TranscriptItem) -> Self {
        match item {
            TranscriptItem::User { text } => {
                bend_engine::AgentMessage::Llm(bend_engine::Message::user(text.clone()))
            }
            TranscriptItem::Assistant {
                text,
                thinking,
                tool_calls,
            } => {
                let mut content = Vec::new();

                if let Some(thinking) = thinking {
                    content.push(bend_engine::Content::Thinking {
                        thinking: thinking.clone(),
                        signature: None,
                    });
                }
                if !text.is_empty() {
                    content.push(bend_engine::Content::Text { text: text.clone() });
                }
                for tool_call in tool_calls {
                    content.push(bend_engine::Content::ToolCall {
                        id: tool_call.id.clone(),
                        name: tool_call.name.clone(),
                        arguments: tool_call.input.clone(),
                    });
                }

                bend_engine::AgentMessage::Llm(bend_engine::Message::Assistant {
                    content,
                    stop_reason: bend_engine::StopReason::Stop,
                    model: String::new(),
                    provider: String::new(),
                    usage: bend_engine::Usage::default(),
                    timestamp: bend_engine::types::now_ms(),
                    error_message: None,
                })
            }
            TranscriptItem::ToolResult {
                tool_call_id,
                tool_name,
                content,
                is_error,
            } => bend_engine::AgentMessage::Llm(bend_engine::Message::ToolResult {
                tool_call_id: tool_call_id.clone(),
                tool_name: tool_name.clone(),
                content: vec![bend_engine::Content::Text {
                    text: content.clone(),
                }],
                is_error: *is_error,
                timestamp: bend_engine::types::now_ms(),
            }),
            TranscriptItem::System { text } => bend_engine::AgentMessage::Extension(
                bend_engine::ExtensionMessage::new("system", serde_json::json!({ "text": text })),
            ),
            TranscriptItem::Extension { kind, data } => bend_engine::AgentMessage::Extension(
                bend_engine::ExtensionMessage::new(kind.clone(), data.clone()),
            ),
        }
    }
}

File: cli/src/session/transcript.ts
/**
 * Parse raw transcript items (from NAPI loadTranscript) into UIMessages
 * with ordered assistant content and completed tool results.
 */

import { replayUserText } from './task-notification.js'
import { toolArgsRecord } from '../term/app/tool-args.js'
import type { TranscriptItem } from '../native/index.js'
import type { UIAssistantBlock, UIMessage } from '../term/app/types.js'

// ---------------------------------------------------------------------------
// Raw transcript item shapes (from Rust TranscriptItem serialization)
// ---------------------------------------------------------------------------

type RawAssistantBlock =
  | { type: 'text'; text: string }
  | { type: 'thinking'; text: string }
  | { type: 'tool_call'; id: string; name: string; input: unknown }

interface RawItem {
  type: string
  // User
  text?: string
  // Assistant content uses an array; tool results use a string.
  content?: RawAssistantBlock[] | string
  // Transcripts written before the canonical content migration.
  content_blocks?: RawAssistantBlock[]
  stop_reason?: string
  // ToolResult
  tool_call_id?: string
  tool_name?: string
  is_error?: boolean
  details?: unknown
  // Stats
  kind?: string
  data?: Record<string, unknown>
  // Compact
  reason?: 'threshold' | 'overflow' | 'manual'
  summary?: string
  tokens_before?: number
  tokens_after?: number
  messages_before?: number
  messages_after?: number
}

// ---------------------------------------------------------------------------
// Main conversion
// ---------------------------------------------------------------------------

export function transcriptToMessages(items: TranscriptItem[]): UIMessage[] {
  const rawItems = items as unknown as RawItem[]
  const messages: UIMessage[] = []
  const toolResults = collectToolResults(rawItems)
  let idx = 0

  for (const item of rawItems) {
    const t = item.type
    if (t === 'user') {
      for (const part of replayUserText(item.text ?? '')) {
        messages.push({
          id: `transcript-${part.kind}-${idx++}`,
          role: part.kind === 'user' ? 'user' : 'system',
          text: part.text,
          timestamp: 0,
        })
      }
    } else if (t === 'assistant') {
      const canonical = Array.isArray(item.content) ? item.content : undefined
      const content = buildAssistantContent(canonical ?? item.content_blocks ?? [], toolResults)
      const text = content
        .filter((block): block is Extract<UIAssistantBlock, { type: 'text' }> => block.type === 'text')
        .map(block => block.text)
        .join('')

      messages.push({
        id: `transcript-assistant-${idx++}`,
        role: 'assistant',
        text,
        timestamp: 0,
        content,
      })
    } else if (t === 'compact') {
      const details = item.details && typeof item.details === 'object' && !Array.isArray(item.details)
        ? item.details as Record<string, unknown>
        : undefined
      messages.push({
        id: `transcript-compact-${idx++}`,
        role: 'assistant',
        text: '',
        timestamp: 0,
        compaction: {
          reason: item.reason ?? 'threshold',
          summary: item.summary ?? '',
          tokensBefore: item.tokens_before ?? 0,
          tokensAfter: item.tokens_after ?? 0,
          messagesBefore: item.messages_before ?? 0,
          messagesAfter: item.messages_after ?? 0,
          method: typeof details?.method === 'string' ? details.method : undefined,
          remoteBlobBytes: typeof details?.remote_blob_bytes === 'number' ? details.remote_blob_bytes : undefined,
          fallbackReason: typeof details?.fallback_reason === 'string' ? details.fallback_reason : undefined,
        },
      })
    }
    // stats, tool_result, system, extension, marker — skipped
  }

  return messages
}

function buildAssistantContent(
  persisted: RawAssistantBlock[],
  toolResults: Map<string, { content: string; isError: boolean; details?: unknown }>,
): UIAssistantBlock[] {
  return persisted.map((block, contentIndex): UIAssistantBlock => {
    switch (block.type) {
      case 'thinking':
        return { type: 'thinking', contentIndex, text: block.text }
      case 'tool_call': {
        const result = toolResults.get(block.id)
        const args = toolArgsRecord(block.input) ?? {}
        return {
          type: 'tool_call',
          contentIndex,
          toolCall: {
            id: block.id,
            name: block.name,
            args,
            status: result ? (result.isError ? 'error' : 'done') : 'queued',
            result: result?.content,
            details: result?.details,
            previewCommand: inferPreviewCommand(block.name, args),
          },
        }
      }
      case 'text':
        return { type: 'text', contentIndex, text: block.text }
    }
  })
}

// ---------------------------------------------------------------------------
// Tool results
// ---------------------------------------------------------------------------

function collectToolResults(items: RawItem[]): Map<string, { content: string; isError: boolean; details?: unknown }> {
  const map = new Map<string, { content: string; isError: boolean; details?: unknown }>()
  for (const item of items) {
    if (item.type === 'tool_result' && item.tool_call_id) {
      map.set(item.tool_call_id, {
        content: typeof item.content === 'string' ? item.content : '',
        isError: item.is_error ?? false,
        details: item.details,
      })
    }
  }
  return map
}

/** Re-derive previewCommand from tool name + args (mirrors engine preview_command logic). */
function inferPreviewCommand(name: string, args: Record<string, unknown>): string | undefined {
  const n = name.toLowerCase()
  if (n === 'bash') {
    const cmd = args.command as string | undefined
    return cmd || undefined
  }
  if (n === 'read') {
    const path = args.path as string | undefined
    if (!path) return undefined
    const offset = args.offset as number | undefined
    const limit = args.limit as number | undefined
    if (offset || limit) return `read ${path} [${offset ?? 1}:${(offset ?? 1) + (limit ?? 0) - 1}]`
    return `read ${path}`
  }
  if (n === 'write') {
    const path = args.path as string | undefined
    return path ? `write ${path}` : undefined
  }
  if (n === 'edit') {
    const path = args.path as string | undefined
    const edits = args.edits as unknown[] | undefined
    const count = edits?.length ?? 1
    return path ? `edit ${path} (${count} replacement(s))` : undefined
  }
  return undefined
}
